code/harness/FsSpecific.cpp

```diff
@@ -0,0 +1,189 @@
+#include "FsSpecific.h"
+
+namespace fs_testing {
+
+using std::string;
+
+namespace {
+
+constexpr char kMkfsStart[] = "mkfs -t ";
+constexpr char kFsckCommand[] = "fsck -T -t ";
+
+constexpr char kExt4RemountOpts[] = "errors=remount-ro";
+
+constexpr char kBtrfsFsckCommand[] = "btrfs check ";
+
+constexpr char kXfsFsckCommand[] = "xfs_repair ";
+}
+
+
+FsSpecific* GetFsSpecific(std::string &fs_type) {
+  // TODO(ashmrtn): Find an elegant way to handle errors.
+  if (fs_type.compare(Ext4FsSpecific::kFsType) == 0) {
+    return new Ext4FsSpecific();
+  } else if (fs_type.compare(BtrfsFsSpecific::kFsType) == 0) {
+    return new BtrfsFsSpecific();
+  } else if (fs_type.compare(F2fsFsSpecific::kFsType) == 0) {
+    return new F2fsFsSpecific();
+  } else if (fs_type.compare(XfsFsSpecific::kFsType) == 0) {
+    return new XfsFsSpecific();
+  }
+  return NULL;
+}
+
+/******************************* Ext4 *****************************************/
+// Weird C++11 rule about static constexpr that aren't "simple".
+constexpr char Ext4FsSpecific::kFsType[];
+
+string Ext4FsSpecific::GetMkfsCommand(string &device_path) {
+  return string(kMkfsStart) + Ext4FsSpecific::kFsType + " " +  device_path;
+}
+
+string Ext4FsSpecific::GetPostReplayMntOpts() {
+  return string(kExt4RemountOpts);
+}
+
+string Ext4FsSpecific::GetFsckCommand(const string &fs_path) {
+  return string(kFsckCommand) + kFsType + " " + fs_path + " -- -y";
+}
+
+FileSystemTestResult::ErrorType Ext4FsSpecific::GetFsckReturn(
+    int return_code) {
+  // The following is taken from the specification in man(8) fsck.ext4.
+  if ((return_code & 0x8) || (return_code & 0x10) || (return_code & 0x20) ||
+      // Some sort of fsck error.
+      return_code & 0x80) {
+    return FileSystemTestResult::kCheck;
+  }
+
+  if (return_code & 0x4) {
+    return FileSystemTestResult::kCheckUnfixed;
+  }
+
+  if ((return_code & 0x1) || (return_code & 0x2)) {
+    return FileSystemTestResult::kFixed;
+  }
+
+  if (return_code == 0) {
+    return FileSystemTestResult::kClean;
+  }
+
+  // Default selection so at least something looks wrong.
+  return FileSystemTestResult::kOther;
+}
+
+string Ext4FsSpecific::GetFsTypeString() {
+  return string(Ext4FsSpecific::kFsType);
+}
+
+bool Ext4FsSpecific::AlwaysRunFsck() {
+  return Ext4FsSpecific::kAlwaysRunFsck;
+}
+
+/******************************* Btrfs ****************************************/
+constexpr char BtrfsFsSpecific::kFsType[];
+
+string BtrfsFsSpecific::GetMkfsCommand(string &device_path) {
+  return string(kMkfsStart) + BtrfsFsSpecific::kFsType + " " +  device_path;
+}
+
+string BtrfsFsSpecific::GetPostReplayMntOpts() {
+  return string();
+}
+
+string BtrfsFsSpecific::GetFsckCommand(const string &fs_path) {
+  return string(kBtrfsFsckCommand) + fs_path;
+}
+
+FileSystemTestResult::ErrorType BtrfsFsSpecific::GetFsckReturn(
```

Comment:
The return results for btrfs, f2fs, and xfs will likely need to be combined with stdout to determine if it actually repaired anything. One always returns 0 and the other two return 0/[-]1 to denote simple "I ran" or "I failed for some reason"

Reply:
Agreed. Will this be in a separate PR or are you going to merge it into this one?

Reply:
I think I will do it in another PR. I would like to get this setup so that other people can start to use it